pubsub/gossipsub/README.md

```diff
@@ -1,7 +1,386 @@
+
+The overlay is initially constructed in a random fashion. Whenever a
+peer joins a topic, then it selects `D` peers (in the topic) at random
+and adds them to the mesh, notifying them with a control message. When
+it leaves the topic, it notifies its peers and forgets the mesh for
+the topic.
+
+The mesh is maintained with the following periodic stabilization
+algorithm:
+
+```
+at each peer:
+  loop:
+    if |peers| < D_low:
+       select D - |peers| non-mesh peers at random and add them to the mesh
+    if |peers| > D_high:
+       select |peers| - D mesh peers at random and remove them from the mesh
+    sleep t
+```
+The parameters of the algorithm are `D` which is the target degree,
+and two relaxed degree parameters `D_low` and `D_high` which represent
+admissible mesh degree bounds.
+
+### gossipsub: The gossiping mesh router
+
+The meshsub router offers a baseline construction with good amplification
+control properties, which we augment with _gossip_ about message flow.
+The gossip is emitted to random subsets of peers not in the mesh, similar
+to randomsub, and it allows us to propagate _metadata_ about message flow
+throughout the network. The metadata can be arbitrary, but as a baseline
+we include the message ids of seen messages in the last few seconds.
+The messages are cached, so that peers receiving the gossip can request
+them for transmission with a control message.
+
+The router can use this metadata to improve the mesh, for instance an
+[episub](episub.md) router built on top of gossipsub can create
+epidemic broadcast trees.  Beyond that, the metadata can restart
+message transmission at different points in the overlay to rectify
+downstream message loss. Or it can simply jump hops opportunistically
+and accelerate message transmission for peers who are at some distance
+in the mesh.
+
+Essentially, gossipsub is a blend of meshsub for data and randomsub
+for mesh metadata. It provides bounded degree and amplification factor
+with the meshsub construction and augments it using gossip propagation
+of metadata with the randomsub technique.
+
+
+## The gossipsub protocol
+
+We can now provide a specification of the pubsub protocol by sketching
+out the router implementation.  The router is backwards compatible
+with floodsub, as it accepts floodsub peers and behaves like floodsub
+towards them.
+
+### Control messages
+
+The protocol defines four control messages:
+- `GRAFT`: graft a mesh link; this notifies the peer that it has been added to the local mesh view.
+- `PRUNE`: prune a mesh link; this notifies the peer that it has been removed from the local mesh view.
+- `IHAVE`: gossip; this notifies the peer that the following messages were recently seen and are available on request.
+- `IWANT`: request transmission of messages announced in an `IHAVE` message.
+
+### Router state
+
+The router maintains the following state:
+- `peers`: a set of all known peers; `peers.gossipsub` denotes the gossipsub peers
```

Comment:
IIUC, these are distinct circuits that do not interact with one another, i.e. the gossipsub operations will not interact with the floodsub ones, and viceversa. In fact, these could even be fully distinct sets with no overlapping peers, right?

If that's the case, it would be great to highlight it. Because the processes below do refer to both of them, so it leads to one thinking that they are somehow interrelated. In fact, at first sight I thought the `fanout` and `floodsub` fulfilled similar roles. Been banging my head a little bit 🙃 

On a related note, if two peers expose both gossipsub and floodsub, they'd participate in both circuits?

Reply:
There are no distinct circuits, the same messages propagate across the overlay regardless of whether it's a gossipsub or floodsub peer.
The difference is that there is no control pruning for floodsub peers, we just forward the messages to them.

Reply:
Yeah, sorry. What I meant is that the gossipsub processes don't interact with the floodsub processes directly. The entrypoint to the "circuits" is the same: the incoming message, but they get treated differently by each protocol with no cross-usage of resources.